mp2-v1/src/values_extraction/gadgets/metadata_gadget.rs

```diff
@@ -28,299 +39,667 @@ use rand::{thread_rng, Rng};
+            })
+            .collect::<Vec<ExtractedColumnInfo>>();
+
+        let data_columns = event
+            .data
             .iter()
             .enumerate()
-            .for_each(|(i, t)| pw.set_bool_target(*t, i < columns_metadata.num_extracted_columns));
-        pw.set_target(
-            metadata_target.evm_word,
-            F::from_canonical_u32(columns_metadata.evm_word),
-        );
+            .map(|(j, &offset)| {
+                let input = [
+                    event.address.as_slice(),
+                    event.event_signature.as_slice(),
+                    DATA_PREFIX,
+                    &[j as u8 + 1],
+                ]
+                .concat()
+                .into_iter()
+                .map(F::from_canonical_u8)
+                .collect::<Vec<F>>();
+
+                let data_id = H::hash_no_pad(&input).elements[0].to_canonical_u64();
+                ExtractedColumnInfo::new(extraction_id.as_slice(), data_id, offset, 32, 0)
+            })
+            .collect::<Vec<ExtractedColumnInfo>>();
+
+        let extracted_columns = [topic_columns, data_columns].concat();
+
+        TableMetadata::<MAX_COLUMNS, 2>::new(
+            &[tx_index_input_column, gas_used_index_column],
+            &extracted_columns,
+        )
     }
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-pub(crate) struct MetadataTarget<const MAX_COLUMNS: usize, const MAX_FIELD_PER_EVM: usize> {
+pub(crate) struct TableMetadataTarget<const MAX_COLUMNS: usize, const INPUT_COLUMNS: usize>
+where
+    [(); MAX_COLUMNS - INPUT_COLUMNS]:,
+{
     #[serde(
         serialize_with = "serialize_long_array",
         deserialize_with = "deserialize_long_array"
     )]
-    /// Information about all columns of the table
-    pub(crate) table_info: [ColumnInfoTarget; MAX_COLUMNS],
-    #[serde(
-        serialize_with = "serialize_array",
-        deserialize_with = "deserialize_array"
-    )]
-    /// Boolean flags specifying whether the i-th column is actual or not
-    pub(crate) is_actual_columns: [BoolTarget; MAX_COLUMNS],
+    /// Information about all input columns of the table
+    pub(crate) input_columns: [InputColumnInfoTarget; INPUT_COLUMNS],
     #[serde(
-        serialize_with = "serialize_array",
-        deserialize_with = "deserialize_array"
+        serialize_with = "serialize_long_array",
+        deserialize_with = "deserialize_long_array"
     )]
-    /// Boolean flags specifying whether the i-th field being processed has to be extracted into a column or not
-    pub(crate) is_extracted_columns: [BoolTarget; MAX_COLUMNS],
-    /// EVM word that should be the same for all columns we’re extracting here
-    pub(crate) evm_word: Target,
+    /// Information about all extracted columns of the table
+    pub(crate) extracted_columns: [ExtractedColumnInfoTarget; MAX_COLUMNS - INPUT_COLUMNS],
+    /// The number of actual columns
+    pub(crate) num_actual_columns: Target,
```

Comment:
I think this should be replaced by a vector of flags which are employed as selectors in digest computation, and then the number of columns should be computed in an ad-hoc method using these flags and the number of input columns. Otherwise, if this is provided as a witness, there might be a mismatch between how many columns we actually add to the digests and the actual number of columns